### ISSUE TITLE
Backup policy

### ISSUE THREAD

**Issue by jesperpedersen** [pgmoneta/pgmoneta#115]
Extend the `retention` configuration option with a backup policy.

Default will be kept as

```
retention = 7 # Number of days
```

but be extended to

```
retention = 7, 4, 12, 5
```

which represents number of days, number of weeks (1 per week), number of months (1 per month) and number of years (1 per year). We should likely look for a backup made on a Monday (or the first of the month/year) for each category.

WAL need only to be kept for the number of days defined.

We can likely do some definition validation in case.

**Comment by Jubilee101** [comment]
Hi @jesperpedersen , as I'm working on this issue, I'd like some clarifications:
1. What do you mean by "look for a backup made on a Monday (or the first of the month/year) for each category", I thought we should just caculate the total number of seconds and check for a backup these seconds before.
2. What kind of definition validation are we looking for?
3. So are we only supporting two kinds of input here? Either the user only input days, or they give all four parameters (days, weeks, months, years)?
4. For the sake of simplicity, I want to treat each month to be 30 days and ignore leap year at least in the beginning, I wonder if this is ok.

And here's a todo list, let me know if I missed anything
- [ ] In `pgmoneta.h` change the `retention` field in the `configuration` struct and `server` struct to `retention_days` (by default is 7), `retention_weeks`(by default is 0), `retention_months`(by default is 0) and `retention_years`(by default is 0)
- [ ] In `configuration.c` add a new function to split the input string to assign values to the 4 fields above, replacing the current usage of `as_int`
- [ ] Also modify other functions involving `retention` field in `configuration.c`, `management.c` and `prometheus.c`
- [ ] In `retain_execute()` modify the current implementation to get the new `check_date`

**Comment by jesperpedersen** [comment]
@Jubilee101 

1) There can be multiple backups per day - we should only keep the latest one; for week, month, year. Day is everything.
2) For initial, I would week == 4 && month >= 1 && month <= 12 && year >= 1
3) Yes
4) I think we need to stick with the month / year roll-over, so latest backup from the 1st

Once we get into the week+ stuff we can "assume" that the backup is the latest backup in place, so that can be "rolled-over". However, latest backup from the 1st is "special".

Yeah, testing this is a bit difficult...

**Comment by Jubilee101** [comment]
Thanks @jesperpedersen
> There can be multiple backups per day - we should only keep the latest one; for week, month, year. Day is everything.

Sorry but I'm still not following, it would be best if you could provide a concrete example. What I understand is that if `retention = 2, 4, 8, 5`, we look back at the date 5 yrs, 8 months and 4 weeks before, let's call it A, and find the latest backup on the nearest Monday before A, then delete all the backups before it. And if `retention = 2, 0, 8, 5`, this means the week is unspecified, so A would be the date 8 months and 4 weeks before, and oldest backup we retain is the latest backup on the first day of the month before A. And in either case, we only keep WAL that's less than 2 days old. (This sounds overly and unnecessarily complicated, so I'm pretty sure I didn't get it)

> I would week == 4 && month >= 1 && month <= 12 && year >= 1

Do you mean 1 <= week <= 4 here?

**Comment by jesperpedersen** [comment]
Well, that is a discussion, right ? Is `7, 4` valid ? Ideally, yes. If so, then `7, 4, 8` is valid too and `7, 4, 12, 5`.

It is only backups "above" days that has to "fit" in, but to make it easier for users to understand it is likely best to have a `7, X` requirement. E.g. if weeks are defined then days must be at least 7.

**Comment by jesperpedersen** [comment]
To make it clear -- for the first pull request make it as simple as possible -- having specific requirements is ok. And `7, 4, 12, 5` would fit that

**Comment by Jubilee101** [comment]
> E.g. if weeks are defined then days must be at least 7.

Wait I thought if weeks are defined then days are at most 7? You are speicifying the day within on week right?

I agree for simple starter I can just keep it as either only days or all four numbers, this way we can make the validation simpler. 

But I'm still not getting the Monday/first day rule you mentioned. Let's say the retention sums up to be N days, why not just find the latest backup on the day N days before and delete the rest?

**Comment by jesperpedersen** [comment]
Yeah, days can be `7, X, ...` max.

Well, days is what is controlling what WAL is kept. But, you have to "roll" a backup into a week, or a month, or a year.

Example -

February 27 - March 6: Daily backups
March 1 - Month backup
February 20 - Week backup
February 13 - Week backup
February 6 - Week backup
February 1 - Month backup
January 30 - Week backup
January 1 - Year backup

**Comment by Jubilee101** [comment]
Oh I see, so if `retention = 7, 4, 2, 8`, we keep all the backups and WAL each day within the nearest 7 days, only the latest backup on the first day of each week within the nearest 4 weeks, only the lastest backup on the first day of each month within the nearest 2 months and so on?

**Comment by Jubilee101** [comment]
Got it, I'll work on it right away, thanks for the patience.

**Comment by Jubilee101** [comment]
> I would week == 4 && month >= 1 && month <= 12 && year >= 1

Hi @jesperpedersen Can you explain more about why week should be exactly 4 here?

**Comment by jesperpedersen** [comment]
Well, it depends on the setup... if `month` is defined then `week` should be max 4, otherwise not

**Comment by Jubilee101** [comment]
Could you explain why here? I thought the `week` and `month` retain backups in their own independent cycles, why does it matter if we retain one more week of backup when `month` is defined

**Comment by jesperpedersen** [comment]
I think that is a matter of documentation - so we need something around this. Maybe a new file for `doc/` ?

**Comment by Jubilee101** [comment]
That would be great. Should I draft it? I don't know if I really understand the validation rule here lol.

**Comment by jesperpedersen** [comment]
I think we need additional documentation for this PR

**Comment by Jubilee101** [comment]
Hmm no problem then, I'll start drafting it. How about we make the rule for specified paramters like this:
days >= 7
1 <= weeks <= 4 when months is specified, weeks >= 1 if otherwise
1 <= months <= 12 when years is specified, months >= 1 if otherwise
years >= 1
Unspecified trailing parameters will be set to 0 and not validated
Specified parameters have to start with days and consecutive

**Comment by jesperpedersen** [comment]
The documentation "just" have to reflect the implementation - we can expand on both later

**Comment by Jubilee101** [comment]
Yeah I agree, though this rule is actually quite easy to implement based on current implementation I think, if my implementation is ok that is.
